hudi-common/src/main/java/org/apache/hudi/common/model/DeleteKey.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.model;
+
+import java.util.Objects;
+
+/**
+ * Delete key is a combination of HoodieKey and ordering value.
+ * The key is used for {@link org.apache.hudi.common.table.log.block.HoodieDeleteBlock}
+ * to support per-record deletions. The deletion block is always appended after the data block,
+ * we need to keep the ordering val to combine with the data records when merging, or the data may
+ * be dropped if there are intermediate deletions for the inputs
+ * (a new INSERT comes after a DELETE in one input batch).
+ */
+public class DeleteKey extends HoodieKey {
```

Comment:
Conceptually, `orderingVal` is not part of the key, it is rather part of the payload. I think it would be better to actually approach it that way:

1. Introduce `DeletionRecordPayload` (carrying only value of `orderingVal`)
2. Introduce new `DeleteLogBlock` format version appropriately persisting/reading new payload type

This has following benefits: 
 - Takes care of backward-compatibility (BWC) out of the box
 - Make it compatible with Metadata Table also out of the box

Reply:
`DeletionRecordPayload` is good suggestion, for a new format i would rather left it to other committers.

Reply:
> For BWC, i would expect for other contributors can address this in following PR.

I don't think we can treat BWC as optional here. What do you think @vinothchandar?

Reply:
> I kind of agree that, but that's the re-design/abstraction of the payload functionality and can we address that in separate issue and don't make things too complex here.

I don't think `DeleteKey` is the right abstraction here for the reasons i've referred to above, of which primary ones are 

 - It's non BWC
 - It's introducing new way of deleting records that make COW and MOR diverge

Reply:
> It's non BWC
Already agreed with that.

> It's introducing new way of deleting records that make COW and MOR diverge
Can you just go over with the work flow in detail of COW and MOR combining process first ? The DELETE records encode/decode for MOR table is always there for efficiency. And i didn't introduce new diverge because it is designed there before. This PR only tag the old delete keys with version number and fix the event time semantics, 

if you want to address that the abstraction is not correct, please give a correct abstraction and let's discuss there and i'm glad to see that.

Reply:
> +1 on providing concrete suggestions to move forward.
> 
> @danny0405 The main contention here is that `DeleteKey` is not really a key. we need some cleanup where we pass the orderingVal from the engine API (flink level), all the way down the actual places where I/O happens. Also mandate orderingVal across all the payloads. I was hoping we would do that in the new merge API design. But that's getting delayed now.

Thanks, here is the case, i'm trying to explain the details again

When consuming the CDC source, flink would tag each record with a `RowKind` corresponding to the `HoodieOperation` field of our format.

When user opens the changelog mode, we do not use the DELETE block encode and treat each delete records as normal data payload in order to keep the deleted val to use for retraction in streaming. So the current code can work correctly.

But in the default upsert mode, say user do not want to keep the hoodie operations flags for streaming read, the bug/problem comes in: We use the DELETE block encoding and if there is disorder deletes in the merge handle batch, the data lost happens.

That's why i try to fix that in this PR then. Hope this can give the help for the ideas of designing new merging APIs.

Reply:
> Can you just go over with the work flow in detail of COW and MOR combining process first ? The DELETE records encode/decode for MOR table is always there for efficiency. And i didn't introduce new diverge because it is designed there before. This PR only tag the old delete keys with version number and fix the event time semantics,

Of course I'm aware of the merging process in both COW and MOR. Efficiency of DELETE block is not in the fact that we just store keys for the records we're planning to delete, right? 

> And i didn't introduce new diverge because it is designed there before. This PR only tag the old delete keys with version number and fix the event time semantics,

I've miscommunicated that -- what i meant was that, adding `DeleteKey` makes it further diverge from the COW implementation.

> +1 on providing concrete suggestions to move forward.

I've outlined concrete suggestion in the very first comments -- to fill in `orderingVal` (as well as potentially additional payload in the future) in a proper `RecordPayload` casing, so that we store not only keys of the deleted records but also a payload with meta information (`orderingVal` for now).